README.md

```diff
@@ -356,7 +361,98 @@ $ ./bin/ipc-agent leave-subnet --subnet=/root/t01002
+
+Here's an example of the configuration connecting to the RPC of the first validator, and configuring all the wallets for the validators in the subnet.
+```toml
+[[subnets]]
+id = "/root/t01002"
+gateway_addr = "t064"
+network_name = "test"
+jsonrpc_api_http = "http://127.0.0.1:1240/rpc/v1"
+auth_token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.JTiumQwFIutkTb0gUC5JWTATs-lUvDaopEDE0ewgzLk"
+accounts = ["t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy", "t1cp4q4lqsdhob23ysywffg2tvbmar5cshia4rweq", "t1nv5jrdxk4ljzndaecfjgmu35k6iz54pkufktvua", "t1pr3qzqieikp33pfdaygwkwt5v35a5ljsxsu64xq", "t1pmxk3nhg2u2ge7ioilrk7c5rmdmmbgpazipvzyy"]
+```
+Remember to run `./bin/ipc-agent reload-config` for your agent to pick up the latest changes for the config.
+
+#### Joining the subnet
+All the infrastructure for the subnet is now deployed, and we can join our validators to the subnet. For this, we need to send a `join` command from each of our validators from their validator wallet addresses providing the validators multiaddress. We can get the validator multiaddress from the output of the script we ran to deploy the infrastructure (or by running `eudico mir validator config validator-addr`). These scripts expose through docker the port for the TPC port of the libp2p host of the validator (if there is interest to use the UDP-based multiaddress drop an issue and we can update them accordingly). Hence, we need to choose from the available multiaddresses one that exposes a TCP connection. Finally, our validators are configured so their docker containers interact with each other through ports in the host machine, so instead of exposing an IP address for the multiaddress, the following DNS-based multiaddress needs to be used for all validators: `/dns/host.docker.internal/`.
+
+This is the command that needs to be executed for every validator to join the subnet:
+```bash
+$ ./bin/ipc-agent join-subnet --from=<validator-wallet> --subnet=/root/t01002 --collateral=<amount-collateral> --validator-net-addr="/dns/host.docker.internal/tcp/<VALIDATOR_PORT>/p2p/<VALIDATOR_MULTIADDR>"
+
+# Sample execution for the validator whose logs where shared above for 2FIL collateral
+$ ./bin/ipc-agent join-subnet --from=t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy --subnet=/root/t01002 --collateral=2 --validator-net-addr="/dns/host.docker.internal/tcp/1359/p2p/12D3KooWEJXcSPw6Yv4jDk52xvp2rdeG3J6jCPX9AgBJE2mRCVoR
+```
+Remember doing the above step for the five validators.
+
+#### Mining in subnet
+We have everything in place now to start mining from all the validators. Mining is as simple as running the following script passing the container id for the validators:
+```bash
+$  ./bin/ipc-infra/mine-subnet.sh <node-container-id>
```

Comment:
Failed, just blocks here:

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-infra/mine-subnet.sh 975db017f41c41ca423fa04205add5f89bc28a8c68c4eadac06e044d0e8d3c5f
[*] Starting to mine on validator
2023-04-01T16:13:05.139Z        INFO    mir-validator-cli       mirvalidator/run.go:143 Checking full node sync status
2023/04/01 16:13:05 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size for details.
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:162 Mir libp2p host listening in the following addresses:
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/172.17.0.2/udp/1348/quic
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/127.0.0.1/udp/1348/quic
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/172.17.0.2/tcp/1347
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/127.0.0.1/tcp/1347
2023-04-01T16:13:05.157Z        INFO    mir-validator-cli       mirvalidator/run.go:228 Starting mining with validator  {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.169Z        INFO    mir-consensus   mir/conf_manager.go:183 stored next configuration number not found      {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.169Z        INFO    mir-consensus   mir/conf_manager.go:195 stored executed configuration number not found  {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.169Z        INFO    mir-consensus   mir/conf_manager.go:208 stored reconfiguration votes not found  {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.170Z        INFO    mir-consensus   mir/manager.go:244      Mir manager serve started       {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.170Z        INFO    mir-consensus   mir/manager.go:248      Mir info:
        Network - /root/t01015
        Validator ID - t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q
        Mir peerID - t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q
        Validators - [t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1 t14obklze7ke4ccepzceo5reqb2kexe4xuskprhlq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWAujMYtMdDXQyq7dPMRHT6pzJo7SDpNEyQwDi5YM21xLE t1qaujcu3rs56sxt4f3ikapzknenvajdvfgc7g3gq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWEDveetd6KRgrdjSjTEKXfe54zKNdRv3DAvVJaChMQr3P t1erhas3yxe7njdlhv3zcm4mzk2wlh56vr5l3czdi@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWLVkXDSk6WteGkUsY2Roi7w8yrESLGJm35SuuUZF51Wzf]   {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:220        RestoreState for epoch 0 started        {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:244        RestoreState: next membership size is 5 at epoch 0      {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:276        Snapshot len is zero    {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:279        RestoreState for epoch 0 finished       {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:463        New epoch started: updating 0 to 0      {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:489        New epoch finished: updating 0 to 0     {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
```

Maybe on account of validator addresses -- see above?

Reply:
The bigger issue is that the user gets stuck here too... I assume I have the wrong validator address. How do I fix it?

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-agent leave-subnet --subnet /root/t01015
[2023-04-01T16:20:18Z ERROR ipc_agent::cli::commands] process command: LeaveSubnet(LeaveSubnetArgs { ipc_agent_url: None, from: None, subnet: "/root/t01015" }) failed due to error: json_rpc error: {"code":-1,"data":null,"message":"json_rpc error: {\"code\":1,\"message\":\"GasEstimateMessageGas error: message execution failed: exit 20, reason: message failed with backtrace:\\n00: f01015 (method 79496443) -- caller has no stake in subnet (20)\\n (RetCode=20)\"}"}
```

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-agent set-validator-net-addr
error: the following required arguments were not provided:
  --subnet <SUBNET>
  --validator-net-addr <VALIDATOR_NET_ADDR>

Usage: ipc-agent set-validator-net-addr --subnet <SUBNET> --validator-net-addr <VALIDATOR_NET_ADDR>
```

Now, this will naturally error out without it as my default wallet isn't a validator and that ` from: None` in the error suggests I should be able to pass a `--from=` to say which validator I want to leave (and, indeed, I can, both for leave-subnet and set-validator-net-addr). But the message & usage instructions don't really mention that possibility and I also didn't notice it at first. So, instead, I just tried to join again with each validator, and it seemed to work:

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-agent join-subnet --from=t1erhas3yxe7njdlhv3zcm4mzk2wlh56vr5l3czdi --subnet=/root/t01015 --collateral=1 --validator-net-addr="/ip4/172.17.0.6/tcp/1347/p2p/12D3KooWLVkXDSk6WteGkUsY2Roi7w8yrESLGJm35SuuUZF51Wzf"
[2023-04-01T16:25:58Z INFO  ipc_agent::cli::commands::manager::join] joined subnet: /root/t01015
```

Except that it didn't. It still gets stuck and the validator list by mine-subnet still shows:
```
        Validators - [t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1 t14obklze7ke4ccepzceo5reqb2kexe4xuskprhlq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWAujMYtMdDXQyq7dPMRHT6pzJo7SDpNEyQwDi5YM21xLE t1qaujcu3rs56sxt4f3ikapzknenvajdvfgc7g3gq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWEDveetd6KRgrdjSjTEKXfe54zKNdRv3DAvVJaChMQr3P t1erhas3yxe7njdlhv3zcm4mzk2wlh56vr5l3czdi@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWLVkXDSk6WteGkUsY2Roi7w8yrESLGJm35SuuUZF51Wzf]
```

So then I passed from to `set-validator-net-addr` for each node, and it went through, i.e. now shows the right set. Still doesn't run, though.

So I tried to restart the containers and `set-validator-net-addr` again, but now I just get this

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc_agent set-validator-net-addr --subnet=/root/t01015 --from=t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty --validator-net-addr=/dns/host.docker.internal/tcp/1250/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1
[2023-04-01T16:56:17Z ERROR ipc_agent::cli::commands] process command: SetValidatorNetAddr(SetValidatorNetAddrArgs { ipc_agent_url: None, from: Some("t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty"), subnet: "/root/t01015", validator_net_addr: "/dns/host.docker.internal/tcp/1250/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1" }) failed due to error: No such file or directory (os error 2)
```

Other commands work and the error is very unhelpful. Inspecting the running docker containers also shows errors but not a lot of directions forward, so I guess I'll just have to start from scratch. But no time to do it now :(

Reply:
> Failed, just blocks here:
This is a validator waiting for the rest of validators to come online. Sometimes the first block requires several rounds of reconfiguration to be produced so it may take ~1min with the current block times. If there was an error on the configuration of the validator's addresses you should have started seeing dialing errors. 

That being said, this interaction is SUPER useful, thank you for guiding me through your reasoning. It is a perfect example of how horrible the UX is and how much it needs to be improved. I have some ideas of things we could improve. I'll start tracking issues :pray:

Reply:
So the issue was that I wasn't starting mining in each of the validators, self-misled by a previous discussion on control. Instructions updated slightly.